pkg/ha/state/state.go

```diff
@@ -27,58 +45,116 @@ type StateView struct {
+func NewLease(c client.LeaseClient, cluster, potentialLeader string, minT, maxT, currentTime time.Time) (*Lease, error) {
+	dbState, err := c.UpdateLease(context.Background(), cluster, potentialLeader, minT, maxT)
+	if err != nil {
+		return nil, fmt.Errorf("could not create new lease: %#v", err)
+	}
+	var maxTimeInstance string
+	if maxT.After(dbState.LeaseUntil) {
+		maxTimeInstance = potentialLeader
+	} else {
+		maxTimeInstance = dbState.Leader
+	}
 
-// haLockState represents the current lock holder
-// as reported from the db
-type HALockState struct {
-	Cluster    string
-	Leader     string
-	LeaseStart time.Time
-	LeaseUntil time.Time
+	return &Lease{
+		cluster:               dbState.Cluster,
+		leader:                dbState.Leader,
+		leaseStart:            dbState.LeaseStart,
+		leaseUntil:            dbState.LeaseUntil,
+		maxTimeSeen:           maxT,
+		maxTimeInstance:       maxTimeInstance,
+		maxTimeSeenLeader:     maxT,
+		recentLeaderWriteTime: currentTime,
+	}, nil
 }
 
-func (h *State) UpdateStateFromDB(latestState *HALockState) {
+// UpdateFromDB uses the supplied client to attempt to update the lease for the potentialLeader.
+// It either updates the lease, or a new leader with the assigned lease interval is set, as
+// signified by the database as the source of truth.
+func (h *Lease) UpdateFromDB(c client.LeaseClient, potentialLeader string, minT, maxT time.Time) error {
+	h._mu.RLock()
+	cluster := h.cluster
+	oldLeader := h.leader
+	h._mu.RUnlock()
+	stateFromDB, err := c.UpdateLease(context.Background(), cluster, potentialLeader, minT, maxT)
+	if err != nil {
+		return fmt.Errorf("could not update lease from db: %#v", err)
+	}
 	h._mu.Lock()
 	defer h._mu.Unlock()
-	h.leader = latestState.Leader
-	h.leaseStart = latestState.LeaseStart
-	h.leaseUntil = latestState.LeaseUntil
+	h.leader = stateFromDB.Leader
+	h.leaseStart = stateFromDB.LeaseStart
+	h.leaseUntil = stateFromDB.LeaseUntil
+	go exposeHAStateToMetrics(cluster, oldLeader, stateFromDB.Leader)
+	return nil
 }
 
-func (h *State) UpdateMaxTimeOnZero(maxT time.Time, replicaName string) {
-	if h.maxTimeSeen.IsZero() {
-		h.maxTimeSeen = maxT
-		h.maxTimeInstance = replicaName
+// SetUpdateFromDB explicitly sets the state. To be used when the ha_service performs a leader change.
+// An error is returned as a safecheck if the stateFromDB is for a different cluster.
+func (h *Lease) SetUpdateFromDB(stateFromDB *client.LeaseDBState) error {
+	h._mu.RLock()
+	if h.cluster != stateFromDB.Cluster {
+		errMsg := `Attempt to set lease state for wrong cluster. 
+Lease belongs to cluster [%s], attempted to update with lease for cluster [%s]`
+		return fmt.Errorf(errMsg, h.cluster, stateFromDB.Cluster)
 	}
+	oldLeader := h.leader
+	h._mu.RUnlock()
+	h._mu.Lock()
+	defer h._mu.Unlock()
+	h.leader = stateFromDB.Leader
+	h.leaseStart = stateFromDB.LeaseStart
+	h.leaseUntil = stateFromDB.LeaseUntil
+	go exposeHAStateToMetrics(stateFromDB.Cluster, oldLeader, stateFromDB.Leader)
```

Comment:
exposeHAstate seems like a very simple non-blocking function. I'd run it inline 
```suggestion
	exposeHAStateToMetrics(stateFromDB.Cluster, oldLeader, stateFromDB.Leader)
```